### PR TITLE
feat: add execute method to commands

### PR DESCRIPTION
**Pull request by lukealvoeiro**
# About

Before this PR, commands would lexed in the following way:
```
/file:)("asf) asdf"


```
After this PR, commands are lexed as so:

- Improves lexing of commands, recognizing command values within quote marks to allow folks to add plenty of spaces
- Implements the execute portion of the command
- Adds the surrounding context to the command, in case the command needs it.

### CODE REVIEW

File: src/goose/cli/prompt/create.py [hunk 1]
```diff
@@ -6,10 +6,10 @@
 
 from goose.cli.prompt.completer import GoosePromptCompleter
 from goose.cli.prompt.lexer import PromptLexer
-from goose.command import get_commands
+from goose.command.base import Command
 
 
-def create_prompt() -> PromptSession:
+def create_prompt(commands: dict[str, Command]) -> PromptSession:
```
Comment (lilydelalande) [inline src/goose/cli/prompt/create.py:12]: I know this is an old function but can we get a docstring since it's part of your PR 🙏

Comment (lukealvoeiro) [reply src/goose/cli/prompt/create.py:12]: for sure, good call!

File: src/goose/command/base.py [hunk 2]
```diff
@@ -8,9 +8,20 @@ class Command(ABC):
...
-        """Execute's the command and replaces it with the output."""
+    def execute(self, query: str, surrounding_context: str) -> Optional[str]:
+        """
+        Execute's the command and replaces it with the output.
+
+        Args:
+            query (str): The query to execute.
+            surrounding_context (str): The full user message that the query is a part of.
+        """
         return ""
```
Comment (lilydelalande) [inline src/goose/command/base.py:27]: why does this return empty string?

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 3]
```diff
@@ -1,34 +1,64 @@
...
+        self.text_prompt_session = PromptSession()
 
-    @staticmethod
-    def create_prompt_session() -> "GoosePromptSession":
-        return GoosePromptSession(create_prompt())
+    def get_message_after_commands(self, message: str) -> str:
+        lexer = PromptLexer(command_names=list(self.commands.keys()))
+        doc = Document(message)
+        lines = []
+        for line_num in range(len(doc.lines)):
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:30]:
```suggestion
        # Iterate through each line of the document
        for line_num in range(len(doc.lines)):
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 4]
```diff
@@ -1,34 +1,64 @@
...
+            classes_in_line = lexer.lex_document(doc)(line_num)
+            line_result = []
+            i = 0
+            while i < len(classes_in_line):
+                if classes_in_line[i][0] == "class:command" and i + 1 < len(classes_in_line):
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:36]:
```suggestion
                # If a command is found and it is not the last part of the line
                if classes_in_line[i][0] == "class:command" and i + 1 < len(classes_in_line):
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 5]
```diff
@@ -1,34 +1,64 @@
...
+                    command_name = classes_in_line[i][1].strip("/").strip(":")
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:38]:
```suggestion
                    # Extract the command name, removing any slashes or colons
                    command_name = classes_in_line[i][1].strip("/").strip(":")
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 6]
```diff
@@ -1,34 +1,64 @@
...
+                    command_value = classes_in_line[i + 1][1]
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py]:
```suggestion
                    # Get the value following the command, assumed to be the command's argument
                    command_value = classes_in_line[i + 1][1]
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 7]
```diff
@@ -1,34 +1,64 @@
...
+                    value_after_execution = self.commands[command_name].execute(command_value, message)
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:46]:
```suggestion
                    # Execute the command with the given argument, expecting a return value
                    value_after_execution = self.commands[command_name].execute(command_value, message)
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 8]
```diff
@@ -1,34 +1,64 @@
...
+                    if value_after_execution is None:
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:50]:
```suggestion
                    # Raise an error if the command did not return a value
                    if value_after_execution is None:
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 9]
```diff
@@ -1,34 +1,64 @@
...
+                        raise ValueError(f"Command {command_name} returned None")
+                    line_result.append(value_after_execution)
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:54]:
```suggestion
                    # Append the result of the command execution to the line results
                    line_result.append(value_after_execution)
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 10]
```diff
@@ -1,34 +1,64 @@
...
+                    i += 1
+                elif classes_in_line[i][0] == "class:text":
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:58]:
```suggestion
               # If the part is plain text, just append it to the line results
                elif classes_in_line[i][0] == "class:text":
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 11]
```diff
@@ -1,34 +1,64 @@
...
+                    line_result.append(classes_in_line[i][1])
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:59]:
```suggestion
# Join all processed parts of the current line and add it to the lines list
                    line_result.append(classes_in_line[i][1])
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 12]
```diff
@@ -1,34 +1,64 @@
...
+                i += 1
+            lines.append("".join(line_result))
+        return "\n".join(lines)
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:66]:
```suggestion
        # Join all processed lines into a single string with newline characters and return

        return "\n".join(lines)
```

File: src/goose/cli/prompt/goose_prompt_session.py [hunk 13]
```diff
@@ -1,34 +1,64 @@
...
+        command_plugins = get_commands()
+        for command, command_cls in command_plugins.items():
+            self.commands[command] = command_cls()
+        self.main_prompt_session = create_prompt(self.commands)
+        self.text_prompt_session = PromptSession()
 
-    @staticmethod
-    def create_prompt_session() -> "GoosePromptSession":
-        return GoosePromptSession(create_prompt())
+    def get_message_after_commands(self, message: str) -> str:
```
Comment (lilydelalande) [inline src/goose/cli/prompt/goose_prompt_session.py:25]: I added some comments to make the flow of logic clearer to the uninitiated but feel free to ignore!

File: src/goose/command/base.py (hunk 2 above)
Comment (lukealvoeiro) [reply src/goose/command/base.py:27]: A command should always return a string (at least a empty string - `""`), but folks may build the completions portion of their command before building the execution part! Returning a string here means that folks will have a decent experience building commands (kinda like [here in the base toolkit](https://github.com/square/goose/blob/dd126afa6c73fe01ad4991a6b45a6d027cb02b29/src/goose/toolkit/base.py#L54))

**Comment by lifeizhou-ap** [comment]
I am curious whether there is any real use case to use the command `execute`?   If not, Shall we wait to implement this until we have the real requirements? In this case the implementation will fit better for the requirement.

It would be good to have a clear purpose while we adding code as code maintenance could be a cost.

**Comment by lukealvoeiro** [comment]
@lifeizhou-ap we have the quorum toolkit which will be using this, as well as the file command. given this, that's why the ticket has a high priority on our internal Linear board.

it's also a bit of a catch-22, where if we don't give engineers the ability to run commands, we won't see additional use-cases for it.

File: src/goose/command/base.py (hunk 2 above)
Comment (lilydelalande) [reply src/goose/command/base.py:27]: oh i didn't see this was the base class. thanks!!

**Comment by jtorreggiani** [comment]
Summary of Goose Toolkit Capabilities and Architecture (comment created with assistance from Goose):

Goose Toolkit Capabilities:
1. Plugin-based Architecture: Allows for modular and extensible functionality.
2. AI-Assisted Development: Integrates AI capabilities to aid in various development tasks.
3. Version Control Integration: Seamlessly works with Git and GitHub for collaborative development.
4. Task Automation: Provides tools for automating repetitive development tasks.
5. Code Analysis: Offers capabilities for analyzing and improving code quality.
6. Project Management: Includes features for managing development workflows and tasks.

Toolkit Architecture:
1. Core Engine: The central component that manages plugin execution and coordinates toolkit operations.
2. Plugin System: A flexible system for adding and managing various plugins.
3. AI Integration Layer: Interfaces with AI models to provide intelligent assistance.
4. Version Control Interface: Connects with Git and GitHub for source control operations.
5. Task Runner: Executes automated tasks and manages workflows.
6. Code Parser and Analyzer: Tools for reading, understanding, and suggesting improvements to code.
7. User Interface: Command-line interface for interacting with the toolkit.
8. Configuration Manager: Handles user preferences and toolkit settings.
9. Logging and Monitoring: Tracks toolkit operations and performance.

This architecture enables Goose to provide a comprehensive, AI-enhanced development environment that can adapt to various project needs and development styles. The modular design allows for easy expansion and customization of the toolkit's capabilities.

**Comment by jtorreggiani** [comment]
Toolkit Architecture Overview (based on current directory structure):

1. Core Structure:
   - The toolkit is organized under the 'src/goose' directory.
   - Key components include CLI, command handling, synopsis generation, and various toolkits.

2. Toolkit System:
   - Located in 'src/goose/toolkit/'
   - Includes base toolkit functionality and specialized toolkits like developer, GitHub, JIRA, linting, and reasoning.
   - Utilizes Jinja templates for prompts (src/goose/toolkit/prompts/)

3. Repository Context:
   - Dedicated module for handling repository context (src/goose/toolkit/repo_context/)
   - Includes utilities and Jinja templates for repo context operations

4. Summarization Capabilities:
   - Comprehensive summarization toolkit (src/goose/toolkit/summarization/)
   - Supports summarizing files, projects, and repositories

5. Utility Functions:
   - Centralized utilities in 'src/goose/utils/'
   - Includes functionalities for cost calculation, exchange creation, user interaction, file operations, and shell commands

6. CLI Interface:
   - Robust CLI system (src/goose/cli/)
   - Includes configuration, main entry point, session management, and custom prompt handling

7. Synopsis System:
   - Dedicated synopsis module (src/goose/synopsis/)
   - Includes moderation, planning, summarization, and system-level functionalities

8. Command Handling:
   - Separate module for command operations (src/goose/command/)
   - Includes base command functionality and file-specific commands

While the critical systems thinking toolkit is not directly visible in this structure, it's likely being integrated into the existing architecture, possibly as a new toolkit under 'src/goose/toolkit/' or as an enhancement to existing components.

The overall architecture suggests a modular, extensible system with clear separation of concerns between different functionalities, allowing for easy integration of new toolkits and capabilities.

**Comment by jtorreggiani** [comment]
Toolkit Architecture Overview (Part 1):

1. Core Structure:
   - The toolkit is organized under the 'src/goose' directory.
   - Key components include CLI, command handling, synopsis generation, and various toolkits.

2. Toolkit System:
   - Located in 'src/goose/toolkit/'
   - Includes base toolkit functionality and specialized toolkits like developer, GitHub, JIRA, linting, and reasoning.
   - Utilizes Jinja templates for prompts (src/goose/toolkit/prompts/)

3. Repository Context:
   - Dedicated module for handling repository context (src/goose/toolkit/repo_context/)
   - Includes utilities and Jinja templates for repo context operations

4. Summarization Capabilities:
   - Comprehensive summarization toolkit (src/goose/toolkit/summarization/)
   - Supports summarizing files, projects, and repositories

(Comment created with assistance from Goose)

**Comment by jtorreggiani** [comment]
Toolkit Architecture Overview (Part 2):

5. Utility Functions:
   - Centralized utilities in 'src/goose/utils/'
   - Includes functionalities for cost calculation, exchange creation, user interaction, file operations, and shell commands

6. CLI Interface:
   - Robust CLI system (src/goose/cli/)
   - Includes configuration, main entry point, session management, and custom prompt handling

7. Synopsis System:
   - Dedicated synopsis module (src/goose/synopsis/)
   - Includes moderation, planning, summarization, and system-level functionalities

8. Command Handling:
   - Separate module for command operations (src/goose/command/)
   - Includes base command functionality and file-specific commands

While the critical systems thinking toolkit is not directly visible in this structure, it's likely being integrated into the existing architecture, possibly as a new toolkit under 'src/goose/toolkit/' or as an enhancement to existing components.

The overall architecture suggests a modular, extensible system with clear separation of concerns between different functionalities, allowing for easy integration of new toolkits and capabilities.

(Comment created with assistance from Goose)

**Comment by jtorreggiani** [comment]
Another test comment added by Goose. This is an automated comment for testing purposes.

**Comment by jtorreggiani** [comment]
Sorry for automated comments on this pull request. Goose used the current directory instead of the plugins repo to send these comments.

**Pull request closed without merging**